sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py

```diff
@@ -0,0 +1,170 @@
+
+
+class EventProcessor(object):
+    def __init__(self, eventhub_client: EventHubClient, consumer_group_name: str,
+                 partition_processor_factory: Callable[..., PartitionProcessor],
+                 partition_manager: PartitionManager, **kwargs):
+        """An EventProcessor automatically creates and runs consumers for all partitions of the eventhub.
+
+        It provides the user a convenient way to receive events from multiple partitions and save checkpoints.
+        If multiple EventProcessors are running for an event hub, they will automatically balance loading. This feature
+        won't be availabe until preview 3.
+
+        :param consumer_group_name: the consumer group that is used to receive events
+        from the event hub that the eventhub_client is going to receive events from
+        :param eventhub_client: an instance of azure.eventhub.aio.EventClient object
+        :param partition_processor_callable: a callable that is called to return a PartitionProcessor
+        :param partition_manager: an instance of a PartitionManager implementation
+        :param initial_event_position: the offset to start a partition consumer if the partition has no checkpoint yet
+        """
+        self._consumer_group_name = consumer_group_name
+        self._eventhub_client = eventhub_client
+        self._eventhub_name = eventhub_client.eh_name
+        self._partition_processor_factory = partition_processor_factory
+        self._partition_manager = partition_manager
+        self._initial_event_position = kwargs.get("initial_event_position", "-1")
+        self._max_batch_size = eventhub_client.config.max_batch_size
+        self._receive_timeout = eventhub_client.config.receive_timeout
+        self._tasks: List[asyncio.Task] = []
+        self._instance_id = str(uuid.uuid4())
+        self._partition_ids = None
+
+    async def start(self):
+        """Start the EventProcessor.
+            1. retrieve the partition ids from eventhubs
+            2. claim partition ownership of these partitions.
+            3. repeatedly call EvenHubConsumer.receive() to retrieve events and
+            call user defined PartitionProcessor.process_events()
+        """
+        log.info("EventProcessor %r is being started", self._instance_id)
+        partition_ids = await self._eventhub_client.get_partition_ids()
+        self.partition_ids = partition_ids
+        claimed_list = await self._claim_partitions()
+        await self._start_claimed_partitions(claimed_list)
+
+    async def stop(self):
+        """Stop all the partition consumer
+
+        It sends out a cancellation token to stop all partitions' EventHubConsumer will stop receiving events.
+
+        """
+        for task in self._tasks:
+            task.cancel()
+            # It's not agreed whether a partition manager has method close().
+        log.info("EventProcessor %r has been cancelled", self._instance_id)
+
+    async def _claim_partitions(self):
```

Comment:
This function should be simplified. And I'm not sure if the logic is correct - are you really supposed to pass in existing non-modified ownership entries? 

A bit of list comprehension and `in`.`not in` could probably reduce the amount of code and make it more readable.

Reply:
For preview 2, one EventProcessor claims all partitions. This function is to:
1. get ownership list from the data store via PartitionManager. 
2. compare this ownership list with all partition_ids that this EventHub has. For all partition_ids that don't have an ownership yet, put them to to_claim_list with new claims. For the rest, put the original ownership to_claim_list. This means all partitions will be in the to_claim_list. Some are new, some are existing.
3. claim partitions in to_claim_list, meaning all partitions are claimed.

This should work fine as I tested it. 
It can be made simpler to just simply claim all partitions without the comparing in step 2. But I want to use all methods designed in PartitionManager so I made it more complex. This give me a better understanding of PartitionManager.

For preview 3, we'll change this part as we're introducing partion consumer load balancing.